### ISSUE TITLE
Bazel cache artifact signing

### ISSUE THREAD

**Issue by mihaigalos** [bazelbuild/bazel#17827]
### Description of the feature request:

Please optionally support signed artifacts in Bazel's remote cache.

This allows a consumer to verify the cache has come from a trusted source.

### What underlying problem are you trying to solve with this feature?

The applications we would build with Bazel need to be secure.

This includes securing them against potential threats which come from manipulating the CAS

### Which operating system are you running Bazel on?

Linux

### What is the output of `bazel info release`?

-

### If `bazel info release` returns `development version` or `(@non-git)`, tell us how you built Bazel.

-

### What's the output of `git remote get-url origin; git rev-parse master; git rev-parse HEAD` ?

```text
-
```

### Have you found anything relevant by searching the web?

-

### Any other information, logs, or outputs that you want to share?

-

**Comment by mihaigalos** [comment]
I had a look at the code and imho, rehashing the downloaded file and comparing it to the requested hash is good enough and no additional signing is required.
This is already implemented [here](https://github.com/bazelbuild/bazel/blob/32468442496783e2e3050a1ec49f6bd9ba11250f/src/main/java/com/google/devtools/build/lib/remote/options/RemoteOptions.java#L546).

**Comment by fmeum** [comment]
Isn't it still possible to maliciously modify the action cache? That could trick a client into downloading a CAS entry with a correct checksum but of malicious nature.

**Comment by mihaigalos** [comment]
If I understand it correctly, a check is performed by default on the client side.
Bazel rehashes the received blob from the remote cache and compares it to the requested hash.

**Comment by fmeum** [comment]
An action cache entry essentially maps the description of an action (inputs, command lines, ...) to the hashes of the outputs. A malicious cache could lie and provide hashes of malicious files instead that were not in fact produced by the action.

The Bazel client can't verify that this doesn't happen: Confirming that the outputs are correct requires running the action locally and checking that the hashes of the outputs match, but that completely defeats the purpose of a remote cache. 

If you want assurance that the cache is trustworthy beyond what transport security offers you, you would probably need some kind of cryptographic signature on the AC entries.

**Comment by mihaigalos** [comment]
> The Bazel client can't verify that this doesn't happen.

This is true. But a successful cache injection would need to exactly match the hash for a given artifact.

That is saying a lot, as a potential attacker would likely need to brute-force change/vary something in the input file to get an exact match.

**Comment by fmeum** [comment]
I don't think that's true, but maybe I'm misunderstanding your point: If I want to inject malicious code into a remote cache, I would first upload my malicious binary to the CAS with its proper hash. I would then upload an AC entry for an existing action such as `gcc -o my_harmless_test my_harmless_test.c` with an output file pointing to the CAS entry for the malicious binary.

When your Bazel client builds and runs `my_harmless_test`, it will pick up that AC entry, request the malicious binary from the CAS and run it. There is no need to compute preimages of hashes for this attack to work.

**Comment by mihaigalos** [comment]
If you look at a CAS entry and hash (SHA256) its contents, the hash needs to exactly match the AC entry.

This is the simple check that bazel does on the client side.
